controller/konnect/reconciler_keysetref.go

```diff
@@ -0,0 +1,186 @@
+				// Reset the KeySetID in the status and set the condition to True.
+				key.Status.Konnect.KeySetID = ""
+				if res, err := updateStatusWithCondition(ctx, cl, key,
+					conditions.KeySetRefValidConditionType,
+					metav1.ConditionTrue,
+					conditions.KeySetRefReasonValid,
+					"KeySetRef is nil",
+				); err != nil || !res.IsZero() {
+					return res, fmt.Errorf("failed to update status: %w", err)
+				}
+
+				// Transfer the ownership back to the ControlPlane if it's resolved.
+				cpRef, hasCPRef := getControlPlaneRef(ent).Get()
+				if hasCPRef {
+					cp, err := getCPForRef(ctx, cl, cpRef, key.GetNamespace())
+					if err != nil {
+						return ctrl.Result{}, fmt.Errorf("failed to get ControlPlane: %w", err)
+					}
+					if res, err := passOwnershipExclusivelyTo(ctx, cl, key, cp); err != nil || !res.IsZero() {
+						return res, fmt.Errorf("failed to transfer ownership to ControlPlane: %w", err)
+					}
+				}
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if keySetRef.Type != configurationv1alpha1.KeySetRefNamespacedRef {
+		ctrllog.FromContext(ctx).Error(fmt.Errorf("unsupported KeySet ref type %q", keySetRef.Type), "entity", ent)
+		return ctrl.Result{}, nil
+	}
+
+	keySet := configurationv1alpha1.KongKeySet{}
+	nn := types.NamespacedName{
+		Name:      keySetRef.NamespacedRef.Name,
+		Namespace: ent.GetNamespace(),
+	}
+	if err := cl.Get(ctx, nn, &keySet); err != nil {
+		if res, errStatus := updateStatusWithCondition(
+			ctx, cl, ent,
+			conditions.KeySetRefValidConditionType,
+			metav1.ConditionFalse,
+			conditions.KeySetRefReasonInvalid,
+			err.Error(),
+		); errStatus != nil || !res.IsZero() {
+			return res, errStatus
+		}
+
+		// If the KongKeySet is not found, we don't want to requeue.
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, ReferencedKongKeySetDoesNotExist{
+				Reference: nn,
+				Err:       err,
+			}
+		}
+
+		return ctrl.Result{}, fmt.Errorf("failed getting KongKeySet %s: %w", nn, err)
+	}
+
+	// If referenced KongKeySet is being deleted, return an error so that we can remove the entity from Konnect first.
+	if delTimestamp := keySet.GetDeletionTimestamp(); !delTimestamp.IsZero() {
+		return ctrl.Result{}, ReferencedKongKeySetIsBeingDeleted{
+			Reference:         nn,
+			DeletionTimestamp: delTimestamp.Time,
+		}
+	}
+
+	// Verify that the KongKeySet is programmed.
+	cond, ok := k8sutils.GetCondition(conditions.KonnectEntityProgrammedConditionType, &keySet)
+	if !ok || cond.Status != metav1.ConditionTrue {
+		if res, err := updateStatusWithCondition(
+			ctx, cl, ent,
+			conditions.KeySetRefValidConditionType,
+			metav1.ConditionFalse,
+			conditions.KeySetRefReasonInvalid,
+			fmt.Sprintf("Referenced KongKeySet %s is not programmed yet", nn),
+		); err != nil || !res.IsZero() {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if res, err := passOwnershipExclusivelyTo(ctx, cl, ent, &keySet); err != nil || !res.IsZero() {
```

Comment:
Can you elaborate on why we need this 🤔 ?

Reply:
TIL that as long as an object has at least one owner reference pointing to an existing object, it [will not get garbage collected](https://github.com/kubernetes/kubernetes/blob/b42772c3b1b5464838761372042891b6888fb2af/pkg/controller/garbagecollector/garbagecollector.go#L567) (which makes sense if you think more about it). 

So, in our case that means that if we keep owner references both to the ControlPlane and KeySet and we remove KeySet object, the dependent Key won't get deleted because it will still have a solid reference to the ControlPlane. That's not the expected behavior, as removal of the associated KeySet deletes a Key on the Konnect side by cascading the delete. We need to reflect this in the Kubernetes world by explicitly passing the ownership to the KeySet *only*.

Reply:
Thanks for the explanation 🙇 

Do you mind adding at least some of those details in the code as comments?

Reply:
[68f59b5](https://github.com/Kong/gateway-operator/pull/663/commits/68f59b596746a89b5f3c3e358144ab3b1797711e)